File: src/FluentCards/ContainerBuilder.cs
namespace FluentCards;

/// <summary>
/// Fluent builder for creating Container elements.
/// </summary>
public class ContainerBuilder
{
    private readonly Container _container = new() { Items = new List<AdaptiveElement>() };

    /// <summary>
    /// Sets the unique identifier for the container.
    /// </summary>
    /// <param name="id">The unique identifier.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder WithId(string id)
    {
        _container.Id = id;
        return this;
    }

    /// <summary>
    /// Sets the style of the container.
    /// </summary>
    /// <param name="style">The container style.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder WithStyle(ContainerStyle style)
    {
        _container.Style = style;
        return this;
    }

    /// <summary>
    /// Sets the vertical alignment of content within the container.
    /// </summary>
    /// <param name="alignment">The vertical alignment.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder WithVerticalContentAlignment(VerticalAlignment alignment)
    {
        _container.VerticalContentAlignment = alignment;
        return this;
    }

    /// <summary>
    /// Enables or disables bleed for the container.
    /// </summary>
    /// <param name="bleed">True to enable bleed, false to disable.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder WithBleed(bool bleed = true)
    {
        _container.Bleed = bleed;
        return this;
    }

    /// <summary>
    /// Sets the minimum height of the container.
    /// </summary>
    /// <param name="minHeight">The minimum height.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder WithMinHeight(string minHeight)
    {
        _container.MinHeight = minHeight;
        return this;
    }

    /// <summary>
    /// Configures the background image for the container.
    /// </summary>
    /// <param name="configure">Action to configure the background image.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder WithBackgroundImage(Action<BackgroundImageBuilder> configure)
    {
        var builder = new BackgroundImageBuilder();
        configure(builder);
        _container.BackgroundImage = builder.Build();
        return this;
    }

    /// <summary>
    /// Configures the action to invoke when the container is selected.
    /// </summary>
    /// <param name="configure">Action to configure the select action.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder WithSelectAction(Action<ActionBuilder> configure)
    {
        var builder = new ActionBuilder();
        configure(builder);
        _container.SelectAction = builder.Build();
        return this;
    }

    /// <summary>
    /// Adds a TextBlock to the container.
    /// </summary>
    /// <param name="configure">Action to configure the TextBlock.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder AddTextBlock(Action<TextBlockBuilder> configure)
    {
        var builder = new TextBlockBuilder();
        configure(builder);
        _container.Items!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Adds an Image to the container.
    /// </summary>
    /// <param name="configure">Action to configure the Image.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder AddImage(Action<ImageBuilder> configure)
    {
        var builder = new ImageBuilder();
        configure(builder);
        _container.Items!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Adds a nested Container to the container.
    /// </summary>
    /// <param name="configure">Action to configure the Container.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder AddContainer(Action<ContainerBuilder> configure)
    {
        var builder = new ContainerBuilder();
        configure(builder);
        _container.Items!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Adds a ColumnSet to the container.
    /// </summary>
    /// <param name="configure">Action to configure the ColumnSet.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder AddColumnSet(Action<ColumnSetBuilder> configure)
    {
        var builder = new ColumnSetBuilder();
        configure(builder);
        _container.Items!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Adds a FactSet to the container.
    /// </summary>
    /// <param name="configure">Action to configure the FactSet.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder AddFactSet(Action<FactSetBuilder> configure)
    {
        var builder = new FactSetBuilder();
        configure(builder);
        _container.Items!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Adds a RichTextBlock to the container.
    /// </summary>
    /// <param name="configure">Action to configure the RichTextBlock.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder AddRichTextBlock(Action<RichTextBlockBuilder> configure)
    {
        var builder = new RichTextBlockBuilder();
        configure(builder);
        _container.Items!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Adds an ActionSet to the container.
    /// </summary>
    /// <param name="configure">Action to configure the ActionSet.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder AddActionSet(Action<ActionSetBuilder> configure)
    {
        var builder = new ActionSetBuilder();
        configure(builder);
        _container.Items!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Adds an AdaptiveElement to the container.
    /// </summary>
    /// <param name="element">The element to add.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ContainerBuilder AddElement(AdaptiveElement element)
    {
        _container.Items!.Add(element);
        return this;
    }

    /// <summary>
    /// Builds and returns the configured Container.
    /// </summary>
    /// <returns>The configured Container instance.</returns>
    public Container Build()
    {
        return _container;
    }
}

File: go/fluentcards/container_builder.go
package fluentcards

// ContainerBuilder builds a Container Adaptive Card element.
type ContainerBuilder struct {
	data map[string]any
}

func newContainerBuilder() *ContainerBuilder {
	return &ContainerBuilder{data: map[string]any{"type": "Container", "items": []any{}}}
}

func (b *ContainerBuilder) WithID(id string) *ContainerBuilder {
	b.data["id"] = id
	return b
}

func (b *ContainerBuilder) WithStyle(style ContainerStyle) *ContainerBuilder {
	b.data["style"] = string(style)
	return b
}

func (b *ContainerBuilder) WithVerticalContentAlignment(alignment VerticalAlignment) *ContainerBuilder {
	b.data["verticalContentAlignment"] = string(alignment)
	return b
}

func (b *ContainerBuilder) WithBleed(bleed bool) *ContainerBuilder {
	b.data["bleed"] = bleed
	return b
}

func (b *ContainerBuilder) WithMinHeight(minHeight string) *ContainerBuilder {
	b.data["minHeight"] = minHeight
	return b
}

func (b *ContainerBuilder) WithSpacing(spacing Spacing) *ContainerBuilder {
	b.data["spacing"] = string(spacing)
	return b
}

func (b *ContainerBuilder) WithSeparator(separator bool) *ContainerBuilder {
	b.data["separator"] = separator
	return b
}

func (b *ContainerBuilder) WithIsVisible(isVisible bool) *ContainerBuilder {
	b.data["isVisible"] = isVisible
	return b
}

func (b *ContainerBuilder) WithBackgroundImage(configure func(*BackgroundImageBuilder)) *ContainerBuilder {
	bib := newBackgroundImageBuilder()
	configure(bib)
	b.data["backgroundImage"] = bib.Build()
	return b
}

func (b *ContainerBuilder) WithSelectAction(configure func(*ActionBuilder)) *ContainerBuilder {
	ab := newActionBuilder()
	configure(ab)
	b.data["selectAction"] = ab.Build()
	return b
}

func (b *ContainerBuilder) AddTextBlock(configure func(*TextBlockBuilder)) *ContainerBuilder {
	tb := newTextBlockBuilder()
	configure(tb)
	b.data["items"] = append(b.data["items"].([]any), tb.Build())
	return b
}

func (b *ContainerBuilder) AddImage(configure func(*ImageBuilder)) *ContainerBuilder {
	ib := newImageBuilder()
	configure(ib)
	b.data["items"] = append(b.data["items"].([]any), ib.Build())
	return b
}

func (b *ContainerBuilder) AddContainer(configure func(*ContainerBuilder)) *ContainerBuilder {
	cb := newContainerBuilder()
	configure(cb)
	b.data["items"] = append(b.data["items"].([]any), cb.Build())
	return b
}

func (b *ContainerBuilder) AddColumnSet(configure func(*ColumnSetBuilder)) *ContainerBuilder {
	cs := newColumnSetBuilder()
	configure(cs)
	b.data["items"] = append(b.data["items"].([]any), cs.Build())
	return b
}

func (b *ContainerBuilder) AddFactSet(configure func(*FactSetBuilder)) *ContainerBuilder {
	fs := newFactSetBuilder()
	configure(fs)
	b.data["items"] = append(b.data["items"].([]any), fs.Build())
	return b
}

func (b *ContainerBuilder) AddRichTextBlock(configure func(*RichTextBlockBuilder)) *ContainerBuilder {
	rtb := newRichTextBlockBuilder()
	configure(rtb)
	b.data["items"] = append(b.data["items"].([]any), rtb.Build())
	return b
}

func (b *ContainerBuilder) AddActionSet(configure func(*ActionSetBuilder)) *ContainerBuilder {
	as := newActionSetBuilder()
	configure(as)
	b.data["items"] = append(b.data["items"].([]any), as.Build())
	return b
}

func (b *ContainerBuilder) AddElement(element Card) *ContainerBuilder {
	b.data["items"] = append(b.data["items"].([]any), element)
	return b
}

func (b *ContainerBuilder) Build() Card {
	return b.data
}
